### PR TITLE
Add steps for simplification inside acc script

### PR DESCRIPTION
**Pull request by lnarmour**
This enables calling the optimal simplifying reduction algo from the acc script. The script has the following new usage:
```
$ acc --help
usage:
  acc [-m] [-v1 alpha_file] [-v2 alpha_file] [-o out_dir]
      [-s [--target-complexity] [--num-simplifications] [--try-splitting]]
      [ALPHA_FILE]
options:
    -v1, --gen-v1-from         : Input Alpha file (*.alpha) used to generate makefile,
                                 wrapper, and verification files
    -v2, --gen-v2-from         : Input Alpha file (*.alpha) used to generate main system
    -o,  --out-dir             : Directory in which to place the output files, if the
                                 path does not exist then it will be created (default: .)
    -m,  --make                : Run make to build the generated files. Note, this may
                                 fail if both v1 and v2 files have not been generated.
                                 Cannot be used with -s since multiple versions may be
                                 generated. (default: false)
    -s,  --simplify            : Run the simplifying reductions algorithm on v2 input
                                 (default: false)
         --num-simplifications : Stop after a number of simplifications is found, or the
                                 exploration terminates, whichever comes first (default: 1)
         --target-complexity   : Target simplified complexity (default: one less than
                                 the input program's complexity)
         --try-splitting       : Consider splits during simplification (default: false)
arguments:
    ALPHA_FILE                 : Input Alpha file used to generate main ystem, makefile,
                                 wrapper, and verification files. (required if neither 
                                 -v1 nor -v2 is specified)
```

## example run

Take the prefix sum:
```
// prefixSum.alpha
affine prefixSum [N] -> {  : N > 0 }
	inputs
		A : {[i0]: 0 <= i0 < N }
	outputs
		C : {[i0]: 0 <= i0 < N }
	when {  : N > 0 } let
		C[i0] = reduce(+, (i,j->i), {: j <= i } : A[j]);
.
```

Run the acc script, specifying to run simplification with `-s`:
```
$ acc -s -o out test.alpha 
[acc]: reading 'test.alpha' file
[acc]: created 'out/simplifications/v0/prefixSum.c' file
[acc]: reading 'test.alpha' file
[acc]: created 'out/prefixSum.ab' file
[acc]: reading 'out/prefixSum.ab' file
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[acc]: created 'out/prefixSum-wrapper.c' file
[acc]: created 'out/prefixSum_verify.c' file
[acc]: created 'out/Makefile' file
```

It creates a separate directory for each distinct simplification, with the simplified *.alpha file (with the path taken to get there) and the *.c file:
```
$ tree out
out
├── Makefile
├── prefixSum-wrapper.c
├── prefixSum.ab
├── prefixSum_verify.c
└── simplifications
    └── v0
        ├── prefixSum.alpha
        └── prefixSum.c
```

Then the simplified C file can simply be moved up two directories and everything can be built as is.
```
$ cd out
$ cp simplifications/v0/prefixSum.c .
$ make
cc prefixSum.c -o prefixSum.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc prefixSum-wrapper.c -o prefixSum prefixSum.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc prefixSum-wrapper.c -o prefixSum.check prefixSum.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc prefixSum_verify.c -o prefixSum_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc prefixSum-wrapper.c -o prefixSum.verify prefixSum.o  prefixSum_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc prefixSum-wrapper.c -o prefixSum.verify-rand prefixSum.o  prefixSum_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
$
$ ./prefixSum.verify-rand 100
Execution time : 0.000014 sec.
TEST for C PASSED
```

### CODE REVIEW

**Comment by lnarmour** [comment]
As a sanity check, confirmed that it reports correct simplifications for the VBI equations too:



<details>
<summary>acc -o out -s --num-simplifications 10 --try-splitting 4d.alpha</summary>

```bash
$ acc -o out -s --num-simplifications 10 --try-splitting 4d.alpha
[acc]: reading '4d.alpha' file
[alpha]: found simplification/v0/energyEquations.alpha
[alpha]: found simplification/v1/energyEquations.alpha
[alpha]: found simplification/v2/energyEquations.alpha
[alpha]: found simplification/v3/energyEquations.alpha
[alpha]: found simplification/v4/energyEquations.alpha
[alpha]: found simplification/v5/energyEquations.alpha
[alpha]: found simplification/v6/energyEquations.alpha
[alpha]: found simplification/v7/energyEquations.alpha
[alpha]: found simplification/v8/energyEquations.alpha
[alpha]: found simplification/v9/energyEquations.alpha
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: created 'out/simplifications/v0/energyEquations.c' file
[acc]: reading '4d.alpha' file
[acc]: created 'out/energyEquations.ab' file
[acc]: reading 'out/energyEquations.ab' file
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[CLooG] INFO: 1 dimensions (over 5) are scalar.
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[CLooG] INFO: 1 dimensions (over 2) are scalar.
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[CLooG] INFO: 1 dimensions (over 3) are scalar.
[acc]: created 'out/energyEquations-wrapper.c' file
[acc]: created 'out/energyEquations_verify.c' file
[acc]: created 'out/Makefile' file
$ cd out/
$ for f in simplifications/*/*.c; do echo $f; cp $f .; make clean; make; energyEquations.verify-rand 100; echo; done;
simplifications/v0/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.046946 sec.
TEST for VBI PASSED

simplifications/v1/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.035387 sec.
TEST for VBI PASSED

simplifications/v2/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.045589 sec.
TEST for VBI PASSED

simplifications/v3/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.045908 sec.
TEST for VBI PASSED

simplifications/v4/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.045166 sec.
TEST for VBI PASSED

simplifications/v5/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.053191 sec.
TEST for VBI PASSED

simplifications/v6/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.053147 sec.
TEST for VBI PASSED

simplifications/v7/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.055215 sec.
TEST for VBI PASSED

simplifications/v8/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.055316 sec.
TEST for VBI PASSED

simplifications/v9/energyEquations.c
rm -f *.o energyEquations energyEquations.check energyEquations.verify energyEquations.verify-rand
cc energyEquations.c -o energyEquations.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm
cc energyEquations-wrapper.c -o energyEquations.check energyEquations.o  -O3  -std=c99  -I/usr/include/malloc/ -lm -DCHECKING
cc energyEquations_verify.c -o energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -c
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
cc energyEquations-wrapper.c -o energyEquations.verify energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY
cc energyEquations-wrapper.c -o energyEquations.verify-rand energyEquations.o  energyEquations_verify.o -O3  -std=c99  -I/usr/include/malloc/ -lm -DVERIFY -DRANDOM
Execution time : 0.054716 sec.
TEST for VBI PASSED
```

</details>

**Comment by rajopadhye** [comment]
This is fantastic.  But  why are there nine outputs from VBI?

**Comment by lnarmour** [comment]
> This is fantastic. But why are there nine outputs from VBI?

Because the `--try-splitting` option was passed. If we don’t try to make splits (i.e., max simplification) then there is only one. But there are many more if splitting is allowed.

Ah OK.  And I think that I know why there is only one without splits.  For reductions whose body has a top level reduction-op (which is commutative) applied to multiple subexpressions,  It's not doing all permutations of the subexpressions (as yet).  If this the case we could get the new one by doing the permutation ourselves in the input program.

File: bundles/alpha.glue.v1/xtend-gen/alpha/glue/v1/.GenerateOldSupportingC.java._trace
Comment (ryanjob42) [inline bundles/alpha.glue.v1/xtend-gen/alpha/glue/v1/.GenerateOldSupportingC.java._trace:1]: Is this trace file used for anything, or can it be removed?

File: bundles/alpha.glue.v2/xtend-gen/alpha/glue/v2/.GenerateNewWriteC.java._trace
Comment (ryanjob42) [inline bundles/alpha.glue.v2/xtend-gen/alpha/glue/v2/.GenerateNewWriteC.java._trace:1]: Same comment as the other trace file. Do we need these?

File: bundles/alpha.glue.v1/xtend-gen/alpha/glue/v1/.GenerateOldSupportingC.java._trace
Comment (lnarmour) [reply bundles/alpha.glue.v1/xtend-gen/alpha/glue/v1/.GenerateOldSupportingC.java._trace:1]: I thought I removed them all, but must have missed these. Good catch, thanks!

commit 42aa899 removes them all

File: bundles/alpha.glue.v2/xtend-gen/alpha/glue/v2/.GenerateNewWriteC.java._trace
Comment (lnarmour) [reply bundles/alpha.glue.v2/xtend-gen/alpha/glue/v2/.GenerateNewWriteC.java._trace:1]: same as above, nope

**Comment by ryanjob42** [comment]
Looks good to me! Everything is working on my end. Once the merge conflicts are fixed, feel free to close the PR whenever.

**Comment by lnarmour** [comment]
> Ah OK. And I think that I know why there is only one without splits. For reductions whose body has a top level reduction-op (which is commutative) applied to multiple subexpressions, It's not doing all permutations of the subexpressions (as yet). If this the case we could get the new one by doing the permutation ourselves in the input program.

@rajopadhye I think you edited my comment instead of quoting the reply in a new comment :). 

Regardless, I believe it is exploring all permutations of subexpressions (as factorization candidates). But, only factoring out the size term leads to a cubic simplification. It’s possible to factor out the asym term, but that path dead-ends.

**Pull request merged**